src/irmin-pack/layered_store.ml

```diff
@@ -200,8 +200,11 @@ struct
     r
 
   (** If the generation changed, then the upper changed too. TODO: This
-      assumption is ok for now, but does not hold if the RW store is opened
-      after the RO or if RW is closed in the meantime. *)
+      assumption is ok for now, but does not hold if:
+
+      - the RW store is opened after the RO,
+      - if RW is closed in the meantime,
+      - if the RW freezes an even number of times before an RO sync. *)
```

Comment:
Good catch, thank you! This is a bug, and we don't have a test for this case either, I'll open a PR to fix this.